File: 12.linked-list.cycle/sol2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function (head) {
  if (head?.next == null) return false;

  let slow = head;
  let fast = head.next;
  while (slow !== fast) {
    if (slow == null || fast == null) return false;

    slow = slow.next;
    fast = fast?.next?.next ?? null;
  }
  return true;
};

File: 12.linked-list.cycle/sol2.cpp
 
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(nullptr) {}
 };
class Solution {
    public:
        bool hasCycle(ListNode *head) {
            if(head == nullptr || head->next == nullptr) return false;
            
            ListNode* slow = head;
            ListNode* fast = head->next;
            while(slow != fast){

                if(slow == nullptr || fast == nullptr) return false;

                slow = slow->next;
                fast = fast->next;
                if(fast != nullptr){
                    fast = fast->next;
                }
                
            }
            return true;
        }
    };